workflow/controller/dag.go

```diff
@@ -380,11 +381,30 @@ func (woc *wfOperationCtx) resolveDependencyReferences(dagCtx *dagContext, task
 	}
 
 	// Perform replacement
+	// Replace Workflow.Spec.Volumes
+	volumes := woc.wf.Spec.Volumes
+	volumesBytes, err := json.Marshal(volumes)
+	if err != nil {
+		return nil, errors.InternalWrapError(err)
+	}
+	fstTmpl := fasttemplate.New(string(volumesBytes), "{{", "}}")
+	newVolumesStr, err := common.Replace(fstTmpl, scope.replaceMap(), true)
+	if err != nil {
+		return nil, err
+	}
+	var newVolumes []apiv1.Volume
+	err = json.Unmarshal([]byte(newVolumesStr), &newVolumes)
+	if err != nil {
+		return nil, errors.InternalWrapError(err)
+	}
+	woc.wf.Spec.Volumes = newVolumes
```

Comment:
It seems like the volume replacement logic is nearly identical in three places. Since all of these are in the same package can we write a helper which is reused here, substituteGlobalsInVolumes, and resolveReferences?

Reply:
Thank you for your review @jessesuen! You are absolutely right about that.

I wrote a helper (`SubstituteParamsInVolumes(params map[string]string)`) which is used in the three places you mentioned. It is a method of `wfOperationCtx` struct. WDYT?